mbus/mbus-tcp.c

```diff
@@ -52,34 +49,34 @@ mbus_tcp_connect(mbus_handle *handle)
         return -1;
 
     host = tcp_data->host;
-    port = tcp_data->port;
+    sprintf(port, "%d", tcp_data->port);
 
-    //
-    // create the TCP connection
-    //
-    if ((handle->fd = socket(AF_INET,SOCK_STREAM, 0)) < 0)
-    {
-        snprintf(error_str, sizeof(error_str), "%s: failed to setup a socket.", __PRETTY_FUNCTION__);
+    memset(&hints, 0, sizeof hints);
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+
+    if ((status = getaddrinfo(host, port, &hints, &servinfo)) != 0) {
+        snprintf(error_str, sizeof(error_str), "%s: getaddrinfo: %s", __PRETTY_FUNCTION__, gai_strerror(status));
         mbus_error_str_set(error_str);
         return -1;
     }
 
-    s.sin_family = AF_INET;
-    s.sin_port = htons(port);
+    // loop through all the results and connect to the first we can
+    for(p = servinfo; p != NULL; p = p->ai_next) {
+        if ((handle->fd = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1) {
+            continue;
+        }
 
-    /* resolve hostname */
-    if ((host_addr = gethostbyname(host)) == NULL)
-    {
-        snprintf(error_str, sizeof(error_str), "%s: unknown host: %s", __PRETTY_FUNCTION__, host);
-        mbus_error_str_set(error_str);
-        return -1;
-    }
+        if (connect(handle->fd, p->ai_addr, p->ai_addrlen) == -1) {
+            close(handle->fd);
+            continue;
+        }
 
-    memcpy((void *)(&s.sin_addr), (void *)(host_addr->h_addr), host_addr->h_length);
+        break; // if we get here, we must have connected successfully
+    }
 
-    if (connect(handle->fd, (struct sockaddr *)&s, sizeof(s)) < 0)
-    {
-        snprintf(error_str, sizeof(error_str), "%s: Failed to establish connection to %s:%d", __PRETTY_FUNCTION__, host, port);
+    if (p == NULL) {
+        snprintf(error_str, sizeof(error_str), "%s: failed to connect", __PRETTY_FUNCTION__);
```

Comment:
Why can't we keep the connection details?

Reply:
This should be just in case connection failed

Reply:
I know, but in this case the connection details are still relevant. Please keep the error behavior as before.

Reply:
Ah I see already, you mean connection details in error message, not about that if NULL... Sorry I still tried to thing what connection details you mean.